Assets/Scripts/AchievementsService.cs

```diff
@@ -348,34 +353,46 @@ public Texture2D GetAchievementLockedIconTexture(string achievementId)
         /// The icon texture associated with the given achievement ID and
         /// determined by the given URI selector.
         /// </returns>
-        private Texture2D GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
+        private Task<Texture2D> GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
         {
-            Texture2D iconTexture = null;
+            TaskCompletionSource<Texture2D> tcs = new();
             foreach (var achievementDef in _achievements)
             {
                 if (achievementDef.AchievementId != achievementId)
                     continue;
 
-                if (_downloadCache.TryGetValue(uriSelector(achievementDef), out byte[] iconBytes) && null != iconBytes)
+                var uri = uriSelector(achievementDef);
+
+                
+                Texture2D textureFromBytes = new(2, 2);
+
+                if (!_downloadCache.ContainsKey(uri))
                 {
-                    Texture2D textureFromBytes = new(2, 2);
-                    if (textureFromBytes.LoadImage(iconBytes))
-                    {
-                        iconTexture = textureFromBytes;
-                    }
-                    else
+                    GetAndCacheData(uri, data =>
                     {
-                        // TODO: Deal with circumstances where image could not be loaded for some reason
-                    }
-                    break;
+                        if (data.result != UnityWebRequest.Result.Success)
+                        {
+                            Debug.Log($"Could not download achievement icon: {data.result}.");
+                            tcs.SetResult(textureFromBytes);
```

Comment:
Should this set an exception, since it resulted in some kind of error?

Reply:
Not all errors / warnings need to throw an exception. As the name implies - exceptions should be used in narrow circumstances (typically in circumstances where the execution of the program cannot continue). In this particular case, the `AchievementService` can continue to function properly even if downloading an individual icon texture fails. 


For posterity and to start conversations in places other than this PR, I have the following additional thoughts:

I admit there is some subjective determination here, and I am not certain I can specifically state where I would draw the line in all circumstances between when to:

- Log a warning.
- Log an error.
- Throw an exception.

Additionally, there is a question about when a log should be made. Should it be logged before the exception is thrown? Should it be thrown by the code that _catches_ the exception? Should different policies apply to warnings vs errors in that respect? These are all questions that are worth answering and considering - questions I don't have definitive answers for - questions for which we have not defined a standard practice within the team.

Reply:
How does the caller know that this failed, and to not use the texture? (same question as above)

Reply:
Through conversations during peer review - we should change this so that when the image fails to load, a `null` texture is returned and a warning is logged.

Reply:
Made changes consistent with aforementioned determination, therefore marking this conversation as resolved.